### PR TITLE
Himbaechel Gowin: HCLK Support

### PR DESCRIPTION
**Pull request by Seyviour**
# HCLK Support

This pull request introduces support for the CLKDIV and CLKDIV2 BELs in Gowin FPGAs. The implementation has been successfully tested on the GW2AR-18 and GW1N-9C models and is expected to be compatible with (most) other Gowin devices as well.

This PR implements a custom placer for CLKDIV/CLKDIV2 (`place_constrained_hclk_cells`) due to the nature of the implicit constraints around placing them. Said placer uses bipartite matching so this PR also adds some code to support that. Every other change is, by my observation, par for the course for adding support for new bels.

Here's a link to the companion PR on Apicula: https://github.com/YosysHQ/apicula/pull/258

### CODE REVIEW

File: himbaechel/uarch/gowin/cst.cc [hunk 1]
```diff
@@ -166,6 +194,27 @@ struct GowinCstReader
                         }
                     }
                 } break;
+                case hclk: {
+                    IdString cell_type = it->second->type;
+                    if (cell_type != IdString(ID_CLKDIV)) {
```
Comment (gatecat) [inline himbaechel/uarch/gowin/cst.cc:199]: `id_CLKDIV` instead of `IdString(ID_CLKDIV)`

File: himbaechel/uarch/gowin/pack.cc [hunk 2]
```diff
@@ -2981,6 +2981,37 @@ struct GowinPacker
...
+    // ===================================
+    void pack_hclk(void)
+    {
+        log_info("Pack HCLK cells...\n");
+
+        for (auto &cell : ctx->cells) {
+            auto ci = cell.second.get();
+            if (ci->type == id_CLKDIV) {
+                NetInfo *in = ci->getPort(id_HCLKIN);
+                CellInfo *this_driver = in->driver.cell;
```
Comment (gatecat) [inline himbaechel/uarch/gowin/pack.cc:2995]: this should probably have some error handling for the case of `HCLKIN` being disconnected/undriven, i.e. either `in` or `in->driver.cell` being `nullptr`

File: himbaechel/uarch/gowin/pack.cc [hunk 3]
```diff
@@ -2981,6 +2981,37 @@ struct GowinPacker
...
+                if (this_driver->type == id_CLKDIV2) {
+                    NetInfo *out = this_driver->getPort(id_CLKOUT);
+                    if (out->users.entries() > 1) {
+                        // We could do as the IDE does sometimes and replicate the CLKDIV2 cell
+                        // as many times as we need. For now, we keep things simple
+                        log_error("CLKDIV2 that drives CLKDIV should drive no other cells");
```
Comment (gatecat) [inline himbaechel/uarch/gowin/pack.cc:3001]: log messages always need to end in `\n`

**Comment by gatecat** [comment]
This all looks reasonable to me - @yrabbit  do you have any thoughts?

**Comment by yrabbit** [comment]
> This all looks reasonable to me - @yrabbit do you have any thoughts?

I don't see anything criminal. It's unusual for me to see something in the prePlace() function, there is probably some philosophical reason for this :)
So this is a good advancement - we will get a simple DVI encoder.

**Comment by Seyviour** [comment]
Thank you, @gatecat. I've added the null checks and corrected the log statement.

**Comment by Seyviour** [comment]
> > This all looks reasonable to me - @yrabbit do you have any thoughts?
> 
> I don't see anything criminal. It's unusual for me to see something in the prePlace() function, there is probably some philosophical reason for this :)
> So this is a good advancement - we will get a simple DVI encoder.

`prePlace` seemed like a fitting home, especially with having to modify the global. The other option was to put it in `pack` 🤔.

Video always feels magical! :)

**Comment by gatecat** [comment]
Should this be merged before or after the apicula PR?

**Comment by Seyviour** [comment]
I think it's best to wait for Apicula. I'll add a comment here when the PR there gets merged.


Thanks for all the work you've put into nextpnr, @gatecat 🙌

File: himbaechel/uarch/gowin/gowin.h [hunk 4]
```diff
@@ -62,6 +62,17 @@ inline bool type_is_dsp(IdString cell_type)
...
 
+// Return true if a cell is CLKDIV
+inline bool type_is_clkdiv(IdString cell_type) { return cell_type == id_CLKDIV; }
+inline bool is_clkdiv(const CellInfo *cell) { return type_is_clkdiv(cell->type); }
+
+// Return true if a cell is CLKDIV2
+inline bool type_is_clkdiv2(IdString cell_type) { return cell_type == id_CLKDIV2; }
+inline bool is_clkdiv2(const CellInfo *cell) { return type_is_clkdiv2(cell->type); }
+
+// Return true for HCLK BELs
```
Comment (yrabbit) [inline himbaechel/uarch/gowin/gowin.h:73]: cells

**Comment by pepijndevos** [comment]
> Should this be merged before or after the apicula PR?

Apicula CI wont pass until this is merged, so I'd prefer to merge this first.

**Comment by Seyviour** [comment]
> There's a bug here that I need to address. Please don't merge yet.

All clear now.

**Comment by pepijndevos** [comment]
Seems like there is a merge conflict

**Comment by Seyviour** [comment]
I missed that somehow. 

> Seems like there is a merge conflict

I've now resolved it

**Comment by gatecat** [comment]
Thanks!

**Comment by Seyviour** [comment]
It's my pleasure. Thanks for merging!

**Pull request merged**